### PR TITLE
feat: add zoom config on UI and fix zoom-related issues

### PR DESCRIPTION
**Pull request by ShGKme**
## ☑️ Resolves

* Fixes: https://github.com/nextcloud/talk-desktop/issues/793
* Adds `zoomFactor` config to application config
* To handle zoom change from different sources, adds listening to config changes feature
* Adds zoom config to the UI
* Considers zoom in **title bar height** and **initial/min window size**

#### Solution complexity

Zoom can be changes with user settings
- Via new user settings
- Via <kbd>CTRL + 🖱️</kbd>/**Touchpad gesture** - implemented by handling the `zoom-changed` event
- Via <kbd>CTRL + +/-</kbd> shortcut - included to the standard corresponding `MenuItem`s and cannot be handled

Because the last option is not possible to handle in Electron at all, and there is no generic zoom changed event, the solution is a bit complex.

#### Zoom step

On each step zoom can be changed:
1. Adding a percent, e.g. `+20%`
   - Scale on level N = 1 + 0.2 * N
   - `20% - 40% - 60% ... 100% - 120% ... 300% - 320%`
   - Simple and understandable for a user
   - Non liner. Change `20% -> 40%` is dramatic `x2` while `300% - 320%` slightly visible
2. Applying a percent, e.g. `x20%`: 
   - Scale on level N = 1.2 ** N
   - `100% - 120% - 144% - 172.8%`
   - Uniform increase
   - Default Electron behavior
   - Hard to understand by user and not beautiful numbers
3. Having a specific pre-defined list of values
   - More complex implementation
   - Beautiful and yet uniform

Option 2 was chosen.

### Settings

settings

### Title bar height change (pay attention on the window buttons)

🏚️ Before | 🏡 After
---|---
before | after

### Windows on 200%

🏚️ Before | 🏡 After
---|---
image | image
image | image
image | image
image | image

### CODE REVIEW

File: src/app/AppConfig.ts [hunk 1]
```diff
@@ -59,19 +64,26 @@ export type AppConfig = {
 	// Nothing yet...
 }
 
+export type AppConfigKey = keyof AppConfig
```
Comment (Antreesy) [inline src/app/AppConfig.ts:67]: doesn't seem to make much difference for me, also TSC doesn't complain without it

File: src/talk/renderer/Settings/DesktopSettingsSection.vue [hunk 2]
```diff
@@ -66,6 +86,40 @@ function relaunch() {
...
+						<template #icon>
+							<IconMinus :size="20" />
+						</template>
+					</NcButton>
+					<NcTextField :id="inputId"
+						class="zoom-input"
+						:aria-describedby="descriptionId"
+						inputmode="number"
+						:value="zoomFactorPercentage"
+						@change="zoomFactorPercentage = $event.target.value" />
```
Comment (Antreesy) [inline src/talk/renderer/Settings/DesktopSettingsSection.vue:109]: Not a big deal, but should we handle border values here (NcTextField, NcButton)?

**Review by Antreesy — APPROVED** [review]
Tested on Win

File: src/talk/renderer/Settings/DesktopSettingsSection.vue (hunk 2 above)
Comment (ShGKme) [reply src/talk/renderer/Settings/DesktopSettingsSection.vue:109]: It is handled by `zoomFactorPercentage`

File: src/app/AppConfig.ts (hunk 1 above)
Comment (ShGKme) [reply src/app/AppConfig.ts:67]: Added usage of it in other places

File: src/talk/renderer/Settings/DesktopSettingsSection.vue (hunk 2 above)
Comment (Antreesy) [reply src/talk/renderer/Settings/DesktopSettingsSection.vue:109]: I meant that it's still possible to set out of range values in UI, they just not kept in config and not reflected in the app window

File: src/talk/renderer/Settings/components/SettingsSelect.vue [hunk 3]
```diff
@@ -33,44 +35,24 @@ export default {
...
+			<slot name="icon" :size="size" />
+		</template>
+
+		<template #default="{ inputId }">
+			<NcSelect v-model="model"
+				class="settings-select"
+				:options="options"
+				:clearable="false"
+				:searchable="false"
+				:input-id="inputId" />
```
Comment (DorraJaouad) [inline src/talk/renderer/Settings/components/SettingsSelect.vue:49]: image

Comment (ShGKme) [reply src/talk/renderer/Settings/components/SettingsSelect.vue]: Fixed

File: src/app/zoom.service.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	// Limit zoom level to custom title bar limits
+	if (zoom < 0.55 || zoom > 5) {
+		return
+	}
+
+	browserWindow.webContents.setZoomFactor(zoom)
+	// Resize the title bar to match the new zoom level
+	browserWindow.setTitleBarOverlay({
```
[the remark this answers is not in the document]
Comment (ShGKme) [reply src/app/zoom.service.ts]: Fixed

File: src/talk/renderer/Settings/DesktopSettingsSection.vue (hunk 2 above)
Comment (ShGKme) [reply src/talk/renderer/Settings/DesktopSettingsSection.vue]: Fixed

Comment (Antreesy) [reply src/talk/renderer/Settings/DesktopSettingsSection.vue]: Yup, works better with manual input. Maybe same getter/setter could be done for `zoomFactor`?
2024-10-29_16h43_11

**Comment by Antreesy** [comment]
After hard restart window size will be 100%, regardless of config value
B | A
-- | --
image | image

Also label next to NcSelect is overlayed

File: src/talk/renderer/Settings/components/SettingsSelect.vue (hunk 3 above)
Comment (Antreesy) [reply src/talk/renderer/Settings/components/SettingsSelect.vue]: image

File: src/talk/renderer/Settings/DesktopSettingsSection.vue (hunk 2 above)
Comment (ShGKme) [reply src/talk/renderer/Settings/DesktopSettingsSection.vue]: It has different limits. Will put it to constants

**Review by DorraJaouad — APPROVED** [review]
Tested ✅ 🦅

**Comment by ShGKme** [comment]
> Also label next to NcSelect is overlayed

This part I cannot reproduce

**Comment by Antreesy** [comment]
> This part I cannot reproduce

Styles aren't scoped to NcDateTimePicker: https://github.com/nextcloud-libraries/nextcloud-vue/pull/6095
image
cc @GretaD

**Review by Antreesy — APPROVED** [review]
So bug is unrelated to this PR, new changes look good

**Comment by ShGKme** [comment]
Squashed fixups, rebased onto main

**Pull request merged**